contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,1427 @@
+            _operator,
+            _amount
+        );
+    }
+
+    /// @notice Returns minimum possible stake for T, KEEP or NU in T denomination
+    function getMinStaked(address _operator, StakingProvider stakingProviders)
+        public
+        view
+        returns (uint96)
+    {
+        OperatorInfo storage operator = operators[_operator];
+        uint256 maxAuthorization = 0;
+        for (uint256 i = 0; i < operator.authorizedApplications.length; i++) {
+            address application = operator.authorizedApplications[i];
+            maxAuthorization = Math.max(
+                maxAuthorization,
+                operator.authorizations[application].authorized
+            );
+        }
+
+        if (maxAuthorization == 0) {
+            return 0;
+        }
+        if (stakingProviders != StakingProvider.T) {
+            maxAuthorization -= Math.min(maxAuthorization, operator.tStake);
+        }
+        if (stakingProviders != StakingProvider.NU) {
+            maxAuthorization -= Math.min(maxAuthorization, operator.nuInTStake);
+        }
+        if (stakingProviders != StakingProvider.KEEP) {
+            maxAuthorization -= Math.min(
+                maxAuthorization,
+                operator.keepInTStake
+            );
+        }
+        return maxAuthorization.toUint96();
+    }
+
+    /// @notice Returns available amount to authorize for the specified application
+    function getAvailableToAuthorize(address _operator, address _application)
+        public
+        view
+        returns (uint96 availableTValue)
+    {
+        OperatorInfo storage operator = operators[_operator];
+        availableTValue =
+            operator.tStake +
+            operator.keepInTStake +
+            operator.nuInTStake;
+        availableTValue -= operator.authorizations[_application].authorized;
+    }
+
+    /// @notice Adds operators to the slashing queue along with the amount.
+    ///         The notifier will receive reward per each operator from
+    ///         notifiers treasury. Can only be called by application
+    ///         authorized for all operators in the array.
+    function notify(
+        uint96 _amount,
+        uint256 _rewardMultiplier,
+        address _notifier,
+        address[] memory _operators
+    ) internal {
+        require(
+            _amount > 0 && _operators.length > 0,
+            "Specify amount and operators to slash"
+        );
+
+        ApplicationInfo storage application = applicationInfo[msg.sender];
+        require(application.approved, "Application is not approved");
+        require(!application.disabled, "Application is disabled");
+
+        for (uint256 i = 0; i < _operators.length; i++) {
+            address operator = _operators[i];
+            require(
+                operators[operator].authorizations[msg.sender].authorized >=
+                    _amount,
+                "Operator didn't authorize sufficient amount to application"
```

Comment:
How about just checking `authorized > 0`? I any case processSlashing will slash only up to authorized and will slash up to 0 in this case. See 591836e

Reply:
I am not sure if `authorized > 0` is enough, I think the problem is that it `require`d. One operator who can not be slashed because it was already slashed to the bone can prevent the entire group from being slashed.

Consider this scenario: My operator 0xABC is in two groups and it has just a minimum stake required to participate. My 0xABC is slashed to 0 in one of the groups. Now, none of these groups can't be slashed because 0xABC has `.authorized == 0`.

Reply:
In other words, I think `authorized > 0` is fine but we should skip instead of reverting. Unless I missed something... 🤔

Reply:
slashing will be outside of this loop, so if previous `processSlashing` will slash to zero then application already should know about this and not send this operator to contract. Also slashed to 0 is the same situation like never authorized and even skipping looks wrong for me. What do you think?

Reply:
I think we may have a different understanding of how operators are picked for work, I think it's different for PRE and beacon / tBTC.

For beacon / tBTC, we have a pool of operators (like a big bag of operator addresses) from which we are selecting groups of a certain size. Let's say that a group size is 100. There are multiple groups active at the given moment and one operator could be in several groups even if they have just a minimum stake.

In other words, my 0xABC can be in 5 active groups, say G1, G2, G3, G4, G5.

If G5 was slashed and 0xABC'c stake was reduced to 0 (thus their authorization), we still want G1, G2, G3, G4 to be slasheable. 0xABC is still operating in this groups. We can no longer slash 0xABC but we want to slash all other 99 operators. Right now, `notify` (thus `slash` and `seize`) will revert for these groups because 0xABC's authorization is 0.

Reply:
ok, I see. Last question before I'll add this changes: so in app skipping this one with 0 authorization will be not possible or too expensive, right?

Reply:
Too expensive - we do not want to iterate through 100 members to check their stake/authorization before calling slashing. Especially that we need to iterate in `notify` anyway. 

I think these are really rare cases and should never happen but we need to handle them. We might be opening doors for some sophisticated attack if we don't handle it.

Reply:
Changed
✔️